src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,700 @@
+// components of the type graph, which are the minimal sets of
+// mutually-recursive types that must be in the same recursion group with each
+// other for the type section to validate.
+//
+// We must additionally ensure that distinct types remain distinct, which would
+// not be the case if we allowed two separate strongly connected components with
+// the same shape to be rewritten to two copies of the same recursion group.
+// Accidentally merging types would be incorrect because it would change the
+// behavior of casts that would have previously been able to differentiate the
+// types.
+//
+// When multiple strongly connected components have the same shape, first try to
+// differentiate them by permuting their types, which keeps the sizes of the rec
+// groups as small as possible. When there are more strongly connected
+// components that are permutations of one another than there are distinct
+// permutations, fall back to keeping the types distinct by adding distinct
+// brand types to the recursion groups to ensure they have different shapes.
+
+#include "ir/module-utils.h"
+#include "ir/type-updating.h"
+#include "pass.h"
+#include "support/disjoint_sets.h"
+#include "support/strongly_connected_components.h"
+#include "support/topological_orders.h"
+#include "wasm-type-shape.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+// Compute the strongly connected components of the private types, being sure to
+// only consider edges to other private types.
+struct TypeSCCs
+  : SCCs<typename std::vector<HeapType>::const_iterator, TypeSCCs> {
+  std::unordered_set<HeapType> includedTypes;
+  TypeSCCs(const std::vector<HeapType>& types)
+    : SCCs(types.begin(), types.end()),
+      includedTypes(types.cbegin(), types.cend()) {}
+  void pushChildren(HeapType parent) {
+    for (auto child : parent.getReferencedHeapTypes()) {
+      if (includedTypes.count(child)) {
+        push(child);
+      }
+    }
+  }
+};
+
+// After all their distinct permutations have been used, different groups with
+// the same shapes must be differentiated by adding in a "brand" type. Even with
+// a brand mixed in, we might run out of permutations, in which case we need a
+// new brand type. This iterator provides an infinite sequence of possible brand
+// types, prioritizing those with the most compact encoding.
+struct BrandTypeIterator {
+  static constexpr size_t optionCount = 18;
+  static std::array<Field, optionCount> fieldOptions;
+  static void initFieldOptions();
+
+  struct FieldInfo {
+    uint8_t index = 0;
+    bool immutable = false;
+
+    operator Field() const {
+      auto field = fieldOptions[index];
+      if (immutable) {
+        field.mutable_ = Immutable;
+      }
+      return field;
+    }
+
+    bool advance() {
+      if (!immutable) {
+        immutable = true;
+        return true;
+      }
+      immutable = false;
+      index = (index + 1) % optionCount;
+      return index != 0;
+    }
+  };
+
+  bool useArray;
```

Comment:
I can't seem to find where this is initialized?

But, separately, is it worth the complexity? Maybe just using structs is simpler.

Reply:
I would say it's worth the complexity since an array encoding is 25% smaller (if I did the math right) than a singleton struct encoding and I don't think it adds a lot of complexity, but I can easily see the argument the other way as well.